Remove unixtimestamp type

dpsutton opened the pull request:
Remove almost all the remaining usages of the coercions as semantic types.

highlighting one of the commit messages:

`:type/UNIXTimestampSeconds and type/ISO8601DateTimeString out of type`

remove the coercions from the type hierarchy. This brought up a
strange issue that has been present for a while: all liquidbase
migrations run and then all data migrations run. They are not
interleaved. This allows for the following scenario:

- data migration migrates all X values to Y in version 26
- liquibase migration migrates all Y values to Z in version 28

But these are not run in version order, but all liquibase migrations
are run and then all data migrations are run. If you were on an old
version for a while, you could end up with Y values even though the
liquibase migration which comes "after" the data migration turns all Y
values into Z values.

This impacts this change because a data migration in this way:
- liquibase changesets 287, 288, 289, and 290 remove all 'type/UNIX*'
and 'type/ISO8601*' semantic types. These were in 0.39
- data migration `migrate-field-types` added in 0.20.0 was looking for
special_type's of "timestamp_milliseconds" and migrating them to
"type/UNIXTimestampMilliseconds".

Since the liquibase runs before the migrate-field-types, it's possible
for a 'type/UNIX*' semantic type to reappear. And since these were
removed from the type system it would fail validation and blow up. In
this case it actually can't happen since the field its attempting to
migrate (special_type) no longer exists. But since the migrations are
not interleaved this problem still exists.

dpsutton committed:
Remove type/UNIX* and type/ISO8601* from frontend

dpsutton committed:
Set effective-type and coercion strategy when syncing

these values will most likely only be available when using tests, as
metadata is very unlikely to have this. As far as I can tell the
toucannery apparatus is the only bit that has this. Its quite
artificial. I'm not sure if this is a good idea.

dpsutton committed:
:type/UNIXTimestampSeconds and type/ISO8601DateTimeString out of type

remove the coercions from the type hierarchy. This brought up a
strange issue that has been present for a while: all liquidbase
migrations run and then all data migrations run. They are not
interleaved. This allows for the following scenario:

- data migration migrates all X values to Y in version 26
- liquibase migration migrates all Y values to Z in version 28

But these are not run in version order, but all liquibase migrations
are run and then all data migrations are run. If you were on an old
version for a while, you could end up with Y values even though the
liquibase migration which comes "after" the data migration turns all Y
values into Z values.

This impacts this change because a data migration in this way:
- liquibase changesets 287, 288, 289, and 290 remove all 'type/UNIX*'
and 'type/ISO8601*' semantic types. These were in 0.39
- data migration `migrate-field-types` added in 0.20.0 was looking for
special_type's of "timestamp_milliseconds" and migrating them to
"type/UNIXTimestampMilliseconds".

Since the liquibase runs before the migrate-field-types, it's possible
for a 'type/UNIX*' semantic type to reappear. And since these were
removed from the type system it would fail validation and blow up. In
this case it actually can't happen since the field its attempting to
migrate (special_type) no longer exists. But since the migrations are
not interleaved this problem still exists.

dpsutton commented on shared/src/metabase/types.cljc:
```diff
@@ -106,18 +106,6 @@
-(derive :type/ISO8601DateTimeString :type/TemporalString)
-(derive :type/ISO8601DateString :type/TemporalString)
-(derive :type/ISO8601TimeString :type/TemporalString)
-
```
due to the weirdness of the migrations i'm thinking about leaving these in the type hierarchy for a bit but not exposing them to the frontend version of the type hierarchy. Meaning the backend wouldn't blow up and the frontend wouldn't be able to select them as a new choice.

dpsutton committed:
whatever prettier

dpsutton committed:
Appease the machines

dpsutton committed:
Restore the unix and iso semantic types to hierarchy

migration tests still use these
`(impl/test-migrations [283 296] [migrate!] ...)`

A few things in tension: the field requires valid semantic_types,
these _were_ valid semantic_types until v39.

dpsutton committed:
Merge branch 'release-x.39.x' into remove-unixtimestamp-type

camsaul commented:
@dpsutton why don't you just remove the data migration and rewrite it as a Liquibase migration -- or even add it to one of the new ones you created for the type system overhaul? That migration always should have been a Liquibase one in the first place IMO.

camsaul commented on shared/src/metabase/types.cljc:
```diff
@@ -104,6 +104,10 @@
 (derive :type/Instant :type/DateTimeWithLocalTZ)
 
+
+;; TODO: remove these. They shouldn't be allowed as semantic types. But migrations use them targeting version
+;; 0.20. Not clear if we should remove the tests or keep these forever
+
```
I don't want to end up keeping things forever -- our code is complicated enough without a bunch of legacy cruft hanging around. I vote to remove them

camsaul commented on test/metabase/sync/sync_metadata/fields/fetch_metadata_test.clj:
```diff
@@ -39,8 +39,8 @@
               :base-type         :type/BigInteger
-              :effective-type    :type/BigInteger
-              :semantic-type     :type/UNIXTimestampMilliseconds
+              :effective-type    :type/DateTime
```
I think these should technically come back with effective type `:type/Instant` now but it probably makes zero difference here

camsaul reviewed (changes requested):
I'd much rather we just remove the old types entirely and add case statements to deal with pre-20 instances to one of your recent Liquibase migrations to migrate the legacy-legacy types than leave crufty code around forever. I think the duplicate types there would cause confusion that could lead to bugs in the future -- people might not know not to use `type/UNIXTimestampSeconds` and accidentally use it in the future

dpsutton committed:
Remove old "coercion" semantic types

dpsutton committed:
Migrate the v20 version semantic types for unix epoch

dpsutton committed:
Time travelling migrations

these target v20 and v29 which at the time they existed had a column
special type not semantic type. But these run after all of the schema
migrations, not interleaved, so they will have a semantic_type not
special_type when they actually run even though they target v20 or v29
data. strange world

camsaul commented on src/metabase/db/data_migrations.clj:
```diff
@@ -126,9 +126,9 @@
 ;;; +----------------------------------------------------------------------------------------------------------------+
 
-;; this is the old new type system. as of v30 these are migrated from a field named special_type to
+;; this is the old new type system. as of v30 these are migrated from a field named semantic_type to
 ;; semantic_type. These migrations target the world before this migration so they are left as is. The migrations talk
```
They're migrated from `sematic_type` to... `semantic_type`? I think you mean v39 too. Not v30

camsaul reviewed (approved):
This looks good to me for now. I think we should probably go in and convert the rest of that Clojure data migration into a Liquibase migration as a tech debt thing in the future (not as part of this PR) along with any other Clojure  data migrations that we can convert to Liquibase ones. Just so we don't run into order-related weirdness in the future. If it was a Liquibase migration in the first place we wouldn't have to deal with all this drama

dpsutton committed:
add migration's new checksum
